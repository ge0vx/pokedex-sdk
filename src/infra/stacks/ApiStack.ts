import { Stack, StackProps } from "aws-cdk-lib";
import { Cors, LambdaIntegration, ResourceOptions, RestApi } from "aws-cdk-lib/aws-apigateway";
import { Construct } from "constructs";

interface ApiStackProps extends StackProps {
    favPokemonIntegration: LambdaIntegration
}

export class ApiStack extends Stack {

    constructor(scope: Construct, id: string, props?: ApiStackProps){
        super(scope, id, props);

        const api = new RestApi(this, 'favPokemonApi');

        const optionsWithCors: ResourceOptions = {
            defaultCorsPreflightOptions: {
                allowOrigins: Cors.ALL_ORIGINS,
                allowMethods: Cors.ALL_METHODS,
            }
        }

        const favPokemonResource = api.root.addResource('favPokemon', optionsWithCors);
        favPokemonResource.addMethod('GET', props.favPokemonIntegration)
        favPokemonResource.addMethod('POST', props.favPokemonIntegration)
        
    }
}