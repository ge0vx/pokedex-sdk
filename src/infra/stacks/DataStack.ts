import { Stack, StackProps } from "aws-cdk-lib";
import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";
import { Construct } from "constructs";

export class DataStack extends Stack {

    public readonly favPokemonTable: ITable;

    constructor(scope: Construct, id: string, props?: StackProps){
        super(scope, id, props)

        this.favPokemonTable = new Table(this, 'favPokemonTable', {
            partitionKey : {
                name: 'id',
                type: AttributeType.STRING
            },
            tableName : 'favPokemonTable'
        })
    }
}