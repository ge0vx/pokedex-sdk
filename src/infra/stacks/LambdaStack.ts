import { Stack, StackProps } from "aws-cdk-lib";
import { LambdaIntegration } from "aws-cdk-lib/aws-apigateway";
import { ITable } from "aws-cdk-lib/aws-dynamodb";
import { Runtime } from "aws-cdk-lib/aws-lambda";
import { NodejsFunction } from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";
import { join } from "path";
import { Effect, PolicyStatement } from "aws-cdk-lib/aws-iam";

interface LambdaStackProps extends StackProps {
    favPokemonTable: ITable
}

export class LambdaStack extends Stack {

    public readonly favPokemonIntegration: LambdaIntegration;
    public readonly pokemonLamdaIntegration: LambdaIntegration;

    constructor(scope: Construct, id: string, props?: LambdaStackProps){
        super(scope, id, props);

        const runtime = Runtime.NODEJS_18_X;
        const handlerFileName = 'handler';

        const favPokemonLambda = new NodejsFunction(this, 'favPokemonLambda',{
            runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'favPokemon', 'handler.ts'),
            environment: {
                TABLE_NAME: props.favPokemonTable.tableName
            }
        })

        favPokemonLambda.addToRolePolicy(new PolicyStatement({
            effect: Effect.ALLOW,
            resources: [props.favPokemonTable.tableArn],
            actions:[
                'dynamodb:PutItem',
                'dynamodb:Scan'
            ]
        }))

        const pokemonLambda = new NodejsFunction(this, 'PokemonLamda', {
            runtime: runtime,
            handler: handlerFileName,
            entry: join(__dirname, '..', '..', 'services', 'pokemon', 'handler.ts'),
        })
        
        this.favPokemonIntegration = new LambdaIntegration(favPokemonLambda)
        this.pokemonLamdaIntegration = new LambdaIntegration(pokemonLambda)
    }
}