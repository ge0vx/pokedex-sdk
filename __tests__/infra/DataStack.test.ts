import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { DataStack } from "../../src/infra/stacks/DataStack";

describe('Initial test suite', ()=>{

    let dataStackTemplate: Template;

    beforeAll(()=>{
        const testApp = new App({
            outdir: 'cdk.out'
        })

        const dataStack = new DataStack(testApp, 'DataStackFavPokemon');
        dataStackTemplate = Template.fromStack(dataStack);
    })


    test('DynamoDB properties', ()=>{
        dataStackTemplate.hasResourceProperties('AWS::DynamoDB::Table', {
            TableName: "favPokemonTable"
        })
    });

    test('DynamoDB Table snapshot', ()=>{
        const dynamoTable = dataStackTemplate.findResources('AWS::DynamoDB::Table')
        expect(dynamoTable).toMatchSnapshot();
    });

    test('CDK Metadata snapshot', ()=>{
        const cdkMetadata = dataStackTemplate.findResources('AWS::CDK::Metadata')
        expect(cdkMetadata).toMatchSnapshot();
    });
});