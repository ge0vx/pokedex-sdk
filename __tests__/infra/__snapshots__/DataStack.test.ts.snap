// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial test suite CDK Metadata snapshot 1`] = `{}`;

exports[`Initial test suite DynamoDB Table snapshot 1`] = `
{
  "favPokemonTable9241278D": {
    "DeletionPolicy": "Retain",
    "Properties": {
      "AttributeDefinitions": [
        {
          "AttributeName": "id",
          "AttributeType": "S",
        },
      ],
      "KeySchema": [
        {
          "AttributeName": "id",
          "KeyType": "HASH",
        },
      ],
      "ProvisionedThroughput": {
        "ReadCapacityUnits": 5,
        "WriteCapacityUnits": 5,
      },
      "TableName": "favPokemonTable",
    },
    "Type": "AWS::DynamoDB::Table",
    "UpdateReplacePolicy": "Retain",
  },
}
`;
